### ISSUE TITLE
Build with me / Ran with me

### ISSUE THREAD

**Issue by pritoms** [pritoms/Cloned#1]

**Comment by pritoms** [comment]
# All

**Comment by pritoms** [comment]
## Metronome Police or Cop

0, ..., N < K : Future
0, ..., N > D : Branch

**Comment by pritoms** [comment]
## Information, Computation, Measurement, Cloning

The sets {{SIDE, MID}, {MID, SIDE}} forms the basis of information processing.

Construct  {(x, x0) -> {x, x)}
Compute {x -> pi(x)}
Measure {x -> psi[x]}
Copy {(x, x0) -> (y[x], 'x')}

Let x0 be the left side while x being on the right side of the mid. We could say that mid defines the substrate where (x, x) is (side, side).

Permuting x to refer to the counterfactual value from x is like {side, {mid, side}} or {{side, mid}, mid}} whose union will either have mid or side as output.
psi at x could be seen as the function of the mid whose position defines it from the two sides overlapping in.
Copy shows the measurement output where y takes the place of x and x takes y with its label being there instead of x. It is the same as decoding the mid-side in terms of left-right where y at x is right at left's position and the label replacing the right due to leftward shift.

**Comment by pritoms** [comment]
**Programming and Application**

There will be two parts.
- [ ] Agent programming and linked structure,
- [ ] System programming and linked agency.

**Comment by pritoms** [comment]
**Creation of Subsystems**

Sending messages to the core system will use one or more subsystems.

The output of each subsystem is a completion and the response of the core system as a whole is also a completion.

Between each intermediate completions, the prompts to their subsystem agents will be constructed by the core system over several stages.

The instruction given to the core system will be prepared to work with the input-output relations where the equivalent structure within the intermediate computation will be prepared and managed by the core system internally.

Programming includes preparing the prompt and function associated with the completion from prompt given.

Completion is evaluated using the function with output given as input and context matters.

**Personal Policy**

Let's create a function which will prepare all the prompts, core system which is defined for the prompt and callback function which collects the output.

The core system will be defined as a class called `Progression`.

It will have methods for doing things to the class variables.

Class variables will include labeled set of attributes and counterfactual features associated with each label.

Methods uses labels to manage subsystem defined for them. For example one of the labels could be used to create executable objects.

**Executable Object**m

Use the standard python backend but it should develop language using predefined constructors that map inputs to outputs.

Use this language to define system structure.

For example `print(time.time())` : `fetch current time and display` and the interpreter defined on those observables will introduce the python code internally.

A system will specifically instructed to work with a definite set of functions.

Another example could be the following:

_Consider this python code:_

```python
import numpy as np

buffer = np.empty((1000, 1000))
x = np.random.rand(2, 1000)
```

This code shouldn't be required to be produced. Instead there shall be numpy functionalities already pre defined and even composed into macros that will maps the following to the previous:

```
create an empty buffer and call it buffer
buffer has 1000 rows and 1000 columns.

store a randomly initialized matrix of shape (2, 10000) into x
```

The above is still not end-to-end and hence will require storing those local variables in memory but this could be done using abstraction as well.

Map the abstract instructions to the generated python code and map that to an abstract statement outlining what has happened to it.


But usually it is better to avoid detailed programming of tasks in modular manner and instead setup frameworks using custom language that solves a problem directly from abstract description to abstract description holding the information about the process.



**Bootstrapping and Abstraction**

Systems that are designed to work with prompts specific to the application and whose purpose is to construct the intermediate code to cause some other prompts can be made to build a subsystem of its own where it abstracts away the work previously done via chaining the output from the subsystem itself while managing its original organization via transferring to a new system where it governed the process.


```python
class Gemini:
    def __init__(self, generation_config, system_instruction):
        self.system_instruction = system_instruction
        self.generation_config = generation_config
        self.model = genai.GenerativeModel(model_name="models/gemini-1.5-flash", generation_config=self.generation_config, system_instruction=self.system_instruction)
        self.records = []

    def set_configuration(self, temperature=1, top_p=0.95, top_k=64, max_output_tokens=8192, stop_sequences=[], response_mime_type="text/plain"):
        return {
                "temperature": temperature,
                "top_p": top_p,
                "top_k": top_k,
                "max_output_tokens": max_output_tokens,
                "stop_sequences": stop_sequences,
                "response_mime_type": response_mime_type,
            }
        
    def __call__(self, prompt):
        completion = self.model.generate_content(prompt)
        self.records.append({prompt : completion.text})
        return self.records[-1][prompt]

    def view_history(self, index):
        return self.records[index]
```

Use this as base model from now on. The standard progression that I want to build on top of it from what it is now is going to include but not limited to these functionalities below:

- [ ] Preparing a system prompt with configuration appropriate for the task
- [ ] Attach at least one subsystem in addition to the running substrate whereby there would be two levels of processing where the other will be constructed from within the current substrate to work at the approximate level
- [ ] Each system and its subsystems are designed according to the information available as intrinsic attributes and which mandates the configurability and preparing of right instruction to ensure the flow of information across the entire system
- [ ] I have not included any suitable configuration for generating from the base however the right adjustment will be imperative
- [ ] The recording shall be designed to capture more information about the situation, especially for multilevel constitutions

**Comment by pritoms** [comment]
> ```python
> import torch
> import torch.nn as nn
> import torch.nn.functional as F
> from torch.distributions import Bernoulli
> from torch.autograd import Variable
> from itertools import count
> import matplotlib.pyplot as plt
> import numpy as np
> import gym
> import pdb
> 
> 
> class PolicyNet(nn.Module):
>     def __init__(self):
>         super(PolicyNet, self).__init__()
> 
>         self.fc1 = nn.Linear(4, 24)
>         self.fc2 = nn.Linear(24, 36)
>         self.fc3 = nn.Linear(36, 1)  # Prob of Left
> 
>     def forward(self, x):
>         x = F.relu(self.fc1(x))
>         x = F.relu(self.fc2(x))
>         x = F.sigmoid(self.fc3(x))
>         return x
> 
> 
> def main():
> 
>     # Plot duration curve: 
>     # From http://pytorch.org/tutorials/intermediate/reinforcement_q_learning.html
>     episode_durations = []
>     def plot_durations():
>         plt.figure(2)
>         plt.clf()
>         durations_t = torch.FloatTensor(episode_durations)
>         plt.title('Training...')
>         plt.xlabel('Episode')
>         plt.ylabel('Duration')
>         plt.plot(durations_t.numpy())
>         # Take 100 episode averages and plot them too
>         if len(durations_t) >= 100:
>             means = durations_t.unfold(0, 100, 1).mean(1).view(-1)
>             means = torch.cat((torch.zeros(99), means))
>             plt.plot(means.numpy())
> 
>         plt.pause(0.001)  # pause a bit so that plots are updated
> 
>     # Parameters
>     num_episode = 5000
>     batch_size = 5
>     learning_rate = 0.01
>     gamma = 0.99
> 
>     env = gym.make('CartPole-v0')
>     policy_net = PolicyNet()
>     optimizer = torch.optim.RMSprop(policy_net.parameters(), lr=learning_rate)
> 
>     # Batch History
>     state_pool = []
>     action_pool = []
>     reward_pool = []
>     steps = 0
> 
> 
>     for e in range(num_episode):
> 
>         state = env.reset()
>         state = torch.from_numpy(state).float()
>         state = Variable(state)
>         env.render(mode='rgb_array')
> 
>         for t in count():
> 
>             probs = policy_net(state)
>             m = Bernoulli(probs)
>             action = m.sample()
> 
>             action = action.data.numpy().astype(int)[0]
>             next_state, reward, done, _ = env.step(action)
>             env.render(mode='rgb_array')
> 
>             # To mark boundarys between episodes
>             if done:
>                 reward = 0
> 
>             state_pool.append(state)
>             action_pool.append(float(action))
>             reward_pool.append(reward)
> 
>             state = next_state
>             state = torch.from_numpy(state).float()
>             state = Variable(state)
> 
>             steps += 1
> 
>             if done:
>                 episode_durations.append(t + 1)
>                 plot_durations()
>                 break
> 
>         # Update policy
>         if e > 0 and e % batch_size == 0:
> 
>             # Discount reward
>             running_add = 0
>             for i in reversed(range(steps)):
>                 if reward_pool[i] == 0:
>                     running_add = 0
>                 else:
>                     running_add = running_add * gamma + reward_pool[i]
>                     reward_pool[i] = running_add
> 
>             # Normalize reward
>             reward_mean = np.mean(reward_pool)
>             reward_std = np.std(reward_pool)
>             for i in range(steps):
>                 reward_pool[i] = (reward_pool[i] - reward_mean) / reward_std
> 
>             # Gradient Desent
>             optimizer.zero_grad()
> 
>             for i in range(steps):
>                 state = state_pool[i]
>                 action = Variable(torch.FloatTensor([action_pool[i]]))
>                 reward = reward_pool[i]
> 
>                 probs = policy_net(state)
>                 m = Bernoulli(probs)
>                 loss = -m.log_prob(action) * reward  # Negtive score function x reward
>                 loss.backward()
> 
>             optimizer.step()
> 
>             state_pool = []
>             action_pool = []
>             reward_pool = []
>             steps = 0
> 
> 
> if __name__ == '__main__':
>     main()
> ```

Issues:

1. Why do env had to be resettled?
2. I don't want to use gym or any other third party environment.
3. The engagement with the environment is not symmetrical.

**Comment by pritoms** [comment]
> > ```python
> > import torch
> > import torch.nn as nn
> > import torch.nn.functional as F
> > from torch.distributions import Bernoulli
> > from torch.autograd import Variable
> > from itertools import count
> > import matplotlib.pyplot as plt
> > import numpy as np
> > import gym
> > import pdb
> > 
> > 
> > class PolicyNet(nn.Module):
> >     def __init__(self):
> >         super(PolicyNet, self).__init__()
> > 
> >         self.fc1 = nn.Linear(4, 24)
> >         self.fc2 = nn.Linear(24, 36)
> >         self.fc3 = nn.Linear(36, 1)  # Prob of Left
> > 
> >     def forward(self, x):
> >         x = F.relu(self.fc1(x))
> >         x = F.relu(self.fc2(x))
> >         x = F.sigmoid(self.fc3(x))
> >         return x
> > 
> > 
> > def main():
> > 
> >     # Plot duration curve: 
> >     # From http://pytorch.org/tutorials/intermediate/reinforcement_q_learning.html
> >     episode_durations = []
> >     def plot_durations():
> >         plt.figure(2)
> >         plt.clf()
> >         durations_t = torch.FloatTensor(episode_durations)
> >         plt.title('Training...')
> >         plt.xlabel('Episode')
> >         plt.ylabel('Duration')
> >         plt.plot(durations_t.numpy())
> >         # Take 100 episode averages and plot them too
> >         if len(durations_t) >= 100:
> >             means = durations_t.unfold(0, 100, 1).mean(1).view(-1)
> >             means = torch.cat((torch.zeros(99), means))
> >             plt.plot(means.numpy())
> > 
> >         plt.pause(0.001)  # pause a bit so that plots are updated
> > 
> >     # Parameters
> >     num_episode = 5000
> >     batch_size = 5
> >     learning_rate = 0.01
> >     gamma = 0.99
> > 
> >     env = gym.make('CartPole-v0')
> >     policy_net = PolicyNet()
> >     optimizer = torch.optim.RMSprop(policy_net.parameters(), lr=learning_rate)
> > 
> >     # Batch History
> >     state_pool = []
> >     action_pool = []
> >     reward_pool = []
> >     steps = 0
> > 
> > 
> >     for e in range(num_episode):
> > 
> >         state = env.reset()
> >         state = torch.from_numpy(state).float()
> >         state = Variable(state)
> >         env.render(mode='rgb_array')
> > 
> >         for t in count():
> > 
> >             probs = policy_net(state)
> >             m = Bernoulli(probs)
> >             action = m.sample()
> > 
> >             action = action.data.numpy().astype(int)[0]
> >             next_state, reward, done, _ = env.step(action)
> >             env.render(mode='rgb_array')
> > 
> >             # To mark boundarys between episodes
> >             if done:
> >                 reward = 0
> > 
> >             state_pool.append(state)
> >             action_pool.append(float(action))
> >             reward_pool.append(reward)
> > 
> >             state = next_state
> >             state = torch.from_numpy(state).float()
> >             state = Variable(state)
> > 
> >             steps += 1
> > 
> >             if done:
> >                 episode_durations.append(t + 1)
> >                 plot_durations()
> >                 break
> > 
> >         # Update policy
> >         if e > 0 and e % batch_size == 0:
> > 
> >             # Discount reward
> >             running_add = 0
> >             for i in reversed(range(steps)):
> >                 if reward_pool[i] == 0:
> >                     running_add = 0
> >                 else:
> >                     running_add = running_add * gamma + reward_pool[i]
> >                     reward_pool[i] = running_add
> > 
> >             # Normalize reward
> >             reward_mean = np.mean(reward_pool)
> >             reward_std = np.std(reward_pool)
> >             for i in range(steps):
> >                 reward_pool[i] = (reward_pool[i] - reward_mean) / reward_std
> > 
> >             # Gradient Desent
> >             optimizer.zero_grad()
> > 
> >             for i in range(steps):
> >                 state = state_pool[i]
> >                 action = Variable(torch.FloatTensor([action_pool[i]]))
> >                 reward = reward_pool[i]
> > 
> >                 probs = policy_net(state)
> >                 m = Bernoulli(probs)
> >                 loss = -m.log_prob(action) * reward  # Negtive score function x reward
> >                 loss.backward()
> > 
> >             optimizer.step()
> > 
> >             state_pool = []
> >             action_pool = []
> >             reward_pool = []
> >             steps = 0
> > 
> > 
> > if __name__ == '__main__':
> >     main()
> > ```
> 
> Issues:
> 
> 1. Why do env had to be resettled?
> 2. I don't want to use gym or any other third party environment.
> 3. The engagement with the environment is not symmetrical.

4. The agent shall be responsible for copying the information variables in the medium.
5. One agent shall represent the environment to the other.

I am going to introduce the harmonic criteria.

**Comment by pritoms** [comment]
**Harmonic Entropy**

There are two parts:
- Order (_consonance_)
- Disorder (_dissonance_)

The balance between the two sides required in a harmonic structure.

Therefore the new list is:
- [ ] Consonance (**order**)
- [ ] Entropy (**adaptability**)
- [ ] Dissonance (**disorder**)

Constructors are classified under each category!


**Phase Space Construction**

- [ ] Harmony : instrumentations of music
- [ ] Timbre : unique identity space
- [ ] Form : structure of the structure
- [ ] Dynamic : ordinal attributes
- [ ] Melody : step sequences
- [ ] Rhythm : cardinal features
- [ ] Texture : architecture of places
- [ ] Tonality : abstract labeling


New measures:
- [ ] Concordance
- [ ] Discordance






```notes
-1 I think, harmonic entropy is the best criterion
-0 Any level of interpretation must comply with this principle
+0 A structure shall only be defined if its associated members cannot refute the harmonic arrangement
+1 A member becomes integrated in so far it can remain so by itself as a whole
```




**Upon finding the first agent with its constructor, I shall use the superposition of all three of us individually as being intrinsically responsible for the development!**

I shall need only a microphone, recorder, constructor ... abstraction, presence function of the room space.




Write below the local code:

**Comment by pritoms** [comment]
**Replacement for Probability**

Any probability function assigns some numerical measure to a situation given a statistical distribution defining the likelihood of the measuring being true for the mapping of observable quantity to another quantity.

It is insincere attempt at modeling the regularities in nature. We shall emancipate our observations from such kinds of predictive systems and to do this we introduce the following:

- [ ] **mid** : similarity,
- [ ] **side** : difference.

A probability function that maps _x_ given _y_ to _z_ could be defined as:

```unit
x : side[+]
z : side[-]
y : mid[~]

def possibility(input, constructor, target):
    **possible** if _constructor_ for mid(input, output) is possible!

    **possible** if _constructor_ for side(input, output) is also possible!
```


Since the mid-side encoder produces a pair of constructor functions that have one additional copy of the two components, it is a catalyst.

A catalyst participate in the reaction process and remain unchanged in its ability to do it again.

Two additional copies can transform the substrate properties in two stages and for only one of the sum and difference is capable of causing the transfer of necessary properties.

Therefore the probability isn't required for a process that nature allows generic methods of bringing into existence.
